### PR TITLE
CASSANDRA-17112: Publish StrictSerializabilityVerifier so Cassandra may use, and fixed bugs found with the integration

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 1]
```diff
@@ -381,7 +386,7 @@ boolean receiveKnowledgePhasedPredecessors(Step propagate, StrictSerializability
             }
 
             boolean updated = false;
-            for (int key = 0 ; key < maxPredecessors.length ; ++key)
+            for (int key = 0; key < maxPredecessors.length; ++key)
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: unnecessary reformatting?

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 2]
```diff
@@ -444,11 +449,11 @@ boolean receiveUnknownStepPredecessor(UnknownStepHolder unknownStep, StrictSeria
         public String toString()
         {
             return "{key: " + ofKey +
-                   ", peers:" + Arrays.toString(maxPeers) +
-                   ", preds:" + Arrays.toString(maxPredecessors) +
-                   (unknownStepPeers != null ? ", peers?:" + unknownStepPeers : "") +
-                   (unknownStepPredecessors != null ? ", preds?:" + unknownStepPredecessors.values() : "") +
-                   "}";
+                    ", peers:" + Arrays.toString(maxPeers) +
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: unnecessary reformatting?

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 3]
```diff
@@ -545,7 +556,7 @@ public void checkForUnwitnessed(int start)
     }
 
     /**
-     *  The history of observations for a given key, or the set of nodes in the graph of observations for this key.
+     * The history of observations for a given key, or the set of nodes in the graph of observations for this key.
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: doesn't seem necessary?

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 4]
```diff
@@ -28,49 +28,53 @@
...
- *              the length of an observed sequence for a key. This imposes a total causal order for a given register
- *              (sequences [], [1], [1,2] have step indexes of 0, 1 and 2 respectively, with 2 necessarily happening after 1 and 0)
- *  predecessor: a step for a referent key that must occur before the referring step and key.
- *               two kinds: 1) those values for keys [B,C..) at step index i for a read of key A, precede key A's step index i +1
- *                          2) those values for keys [B,C..) at step index i for a write of key A, precede key A's step index i
- *  max predecessor: the maximum predecessor that may be reached via any predecessor relation
- *
+ * register: the values associated with a given key
+ * step: a logical point in the causal sequence of events for a register
+ * step index: the index of a step; since we observe a strictly growing sequence this translates directly to
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: why removing the space at front? it works as indentation under the Nomenclature heading?

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 5]
```diff
@@ -606,7 +623,9 @@ void insert(Step step)
                 steps = Arrays.copyOf(steps, Math.max(i + 1, steps.length * 2));
 
             steps[i] = step;
-            while (--i >= 0 && steps[i] == null) {}
+            while (--i >= 0 && steps[i] == null)
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: unnecessary reformatting

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 6]
```diff
@@ -616,12 +635,13 @@ void insert(Step step)
             else
             {
                 i = step.ofStepIndex;
-                while (++i < steps.length && steps[i] == null) {}
+                while (++i < steps.length && steps[i] == null)
+                {
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: unnecessary reformatting (and this was intentional when written)

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 7]
```diff
@@ -782,7 +802,7 @@ public void apply(int start, int end)
             }
         }
 
-        for (int k = 0; k < bufReads.length ; ++k)
+        for (int k = 0; k < bufReads.length; ++k)
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: unnecessary reformatting

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 8]
```diff
@@ -771,7 +791,7 @@ public void witnessWrite(int key, int id)
     public void apply(int start, int end)
     {
         boolean hasUnknownSteps = false;
-        for (int k = 0; k < bufReads.length ; ++k)
+        for (int k = 0; k < bufReads.length; ++k)
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: more unnecessary reformatting

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 9]
```diff
@@ -562,9 +573,15 @@ public void checkForUnwitnessed(int start)
...
+        void print()
+        {
+            for (Map.Entry<Integer, UnknownStepHolder> e : futureWrites.byWriteValue.entrySet())
+                logger.error("{}: {} -> ({}, {}, {})", key, e.getKey(), e.getValue().writeValue, e.getValue().start, e.getValue().end);
+        }
+
         private void updateSequence(int[] sequence, int maybeWrite)
         {
-            for (int i = 0, max = Math.min(sequence.length, this.sequence.length) ; i < max ; ++i)
+            for (int i = 0, max = Math.min(sequence.length, this.sequence.length); i < max; ++i)
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: more unnecessary reformatting

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 10]
```diff
@@ -476,7 +481,7 @@ class FutureWrites extends Step
         void newSequence(int[] oldSequence, int[] newSequence)
         {
             boolean updated = false;
-            for (int i = oldSequence.length ; i < newSequence.length ; ++i)
+            for (int i = oldSequence.length; i < newSequence.length; ++i)
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: more unnecessary reformatting

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 11]
```diff
@@ -349,7 +354,7 @@ boolean updatePredecessorsOfWrite(int[][] reads, int[] writes, StrictSerializabi
                 return false;
 
             boolean updated = false;
-            for (int key = 0 ; key < reads.length ; ++key)
+            for (int key = 0; key < reads.length; ++key)
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: more unnecessary reformatting

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java [hunk 12]
```diff
@@ -315,7 +320,7 @@ void setSuccessor(Step successor)
         boolean updatePeers(int[] newPeers, UnknownStepHolder[] unknownSteps)
         {
             boolean updated = false;
-            for (int key = 0 ; key < newPeers.length ; ++key)
+            for (int key = 0; key < newPeers.length; ++key)
```
Comment (belliottsmith) [inline accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: more unnecessary reformatting

**Review by belliottsmith — CHANGES_REQUESTED** [review]
Most of this patch seems to be applying your own personal style modifications, that are unnecessary.

**Comment by belliottsmith** [comment]
Also, should improve tests to cover these issues, see my branch below. Also, a minor nit suggestion for SSV itself.

https://github.com/belliottsmith/cassandra-accord/tree/CASSANDRA-17112-suggest

**Comment by dcapwell** [comment]
> Most of this patch seems to be applying your own personal style modifications, that are unnecessary.

The style changes match C* style, so unless Accord has a documented style guide that is different than C*, I am not applying my "own personal style" but rather using the projects...

There are only a small number which are not in our style... they were IntelliJ matching java's... these are the doc space changes...

 > that are unnecessary.
 
 Correct, they are style changes.  We talked about this a long time back; when touching code that doesn't match our style its fine to upgrade it to the current style...

**Comment by dcapwell** [comment]
> Also, should improve tests to cover these issues, see my branch below. Also, a minor nit suggestion for SSV itself.
> 
> https://github.com/belliottsmith/cassandra-accord/tree/CASSANDRA-17112-suggest

Thanks, ill take a look.  There are tests in the C* side that cover what we talked about, but will look closer at your branch as well.

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java (hunk 5 above)
Comment (dcapwell) [reply accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: Seems that this recently changed https://cassandra.apache.org/_/development/code_style.html

> { and } are placed on a new line except when empty or opening a multi-line lambda expression.

I know there was a new vote/discussion on style changes and didn't keep up... will need to review...

for now ill revert

**Comment by belliottsmith** [comment]
> The style changes match C* style, so unless Accord has a documented style guide that is different than C*, I am not applying my "own personal style" but rather using the projects...

They are your personal interpretation. You have imposed four changes:

1) You have introduced new lines for empty loop braces. The existing formulation was explicitly permitted by the style guide.
2) You have modified the spacing for new lines for `toString()` methods. Your spacing is not specified anywhere in the style guide to my knowledge, and in general the multi-line guidance while detailed is explicitly crafted to permit a degree of author flexibility for better layout.
3) Removal of spaces in javadoc. This is not provided for anywhere, that I know of, and again in some cases go explicitly against the author's intent.
4) Removing spacing before the semi-colon in a for-loop guard. This one you have a case for, but is in my opinion not explicitly stated in the style guide, as we only inherit these from the sun style guide which simply says they are of the form `for (expression; expression; expression)`, and since spaces are valid in expressions in Java, this does not prohibit the use of a space before the semi-colon, only enforces it afterwards. This one might be debated to be in your favour, but it is not as clear as you perhaps believe.

> Correct, they are style changes. We talked about this a long time back; when touching code that doesn't match our style its fine to upgrade it to the current style...

From the style guide:

"Note that the project has a variety of styles that have accumulated in different subsystems. Where possible a balance should be struck between these guidelines and the style of the code that is being modified as part of a patch. *Patches should also limit their scope to the minimum necessary for safely addressing the concerns of the patch.*"

If you were rewriting these lines for functional or refactoring reasons, your changes would be fine. But you were not, so they are contrary to the style guide (and, in my humble opinion, etiquette).

File: accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java (hunk 9 above)
Comment (dcapwell) [reply accord-core/src/test/java/accord/verify/StrictSerializabilityVerifier.java]: Our style guide does not cover this (see https://cassandra.apache.org/_/development/code_style.html) and instead delegates to Oracle's (see http://java.sun.com/docs/codeconv/html/CodeConvTOC.doc.html).  Based off Oracle's style guide, we should not have the spaces

https://www.oracle.com/java/technologies/javase/codeconventions-statements.html#454

**Comment by dcapwell** [comment]
We spoke in slack, I reverted all style changes

**Review by belliottsmith — APPROVED** [review]

**Comment by belliottsmith** [comment]
+1

**Pull request merged**